### PR TITLE
Add the PgSQL driver

### PR DESCRIPTION
**Pull request by derrabus**
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | N/A

#### Summary

This PR introduces a new Postgres driver that leverages PHP's `pgsql` extension. Postgres is currently the last database that we only support through PDO. This driver closes the gap.

`ext-pgsql` has a rather verbose API. Unlike other database extensions, it does not even bother to cast result types: It returns everything as a string. Except for `null` values: Those are really `null`!

However, we get information on the database type of all result fields, so we are able to perform those casts ourselves. This is why I decided to perform very basic type mappings in the driver already. Otherwise, the result sets produced by `ext-pgsql` would have been a permanent edge case: Booleans come as `'t'` and `'f'`, binary data is returned as hex string prefixed with `0x`. Not casting those would cause a lot of functional tests to fail.

The type casts that I perform in the driver should be similar to what `PDO_pgsql` does internally as well. That should make a potential switch from PDO to the native extension rather seamless.

In PHP 8.1, all resource types used by the extensions have been changed to objects. This is why you will see a couple of type annotations with a union of `resource` and some class from the `PgSql` namespace. This PR currently targets DBAL 3.6 and PHP 7.4. Once this code hits the 4.x branch, we can simplify a lot of type declarations.

The one big limitation though is that I haven't implemented support for loading BLOBs from resources yet. That topic is a bit complex, so I'd like to work on that in a follow-up. For now, I think we can ship this driver without BLOB support.

### CODE REVIEW

File: psalm.xml.dist [hunk 1]
```diff
@@ -19,7 +19,6 @@
         <file name="vendor/jetbrains/phpstorm-stubs/ibm_db2/ibm_db2.php" />
         <file name="vendor/jetbrains/phpstorm-stubs/mysqli/mysqli.php" />
         <file name="vendor/jetbrains/phpstorm-stubs/oci8/oci8.php" />
-        <file name="vendor/jetbrains/phpstorm-stubs/pgsql/pgsql.php" />
```
Comment (derrabus) [inline psalm.xml.dist:22]: Psalm actually runs better using its own `ext-pgsql` stubs. 🧐

File: src/Driver/PgSQL/Connection.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    private $connection;
+
+    private Parser $parser;
+
+    /** @param PgSqlConnection|resource $connection */
+    public function __construct($connection)
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource of an instance of %s, got %s.',
```
Comment (greg0ire) [inline src/Driver/PgSQL/Connection.php]:
```suggestion
                'Expected connection to be a resource or an instance of %s, got %s.',
```

File: src/Driver/PgSQL/Connection.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                'The usage of Connection::lastInsertId() with a sequence name is deprecated.',
+            );
+
+            return $this->query(sprintf('SELECT CURRVAL(%s)', $this->quote($name)))->fetchOne();
+        }
+
+        return $this->query('SELECT LASTVAL()')->fetchOne();
+    }
+
+    public function beginTransaction(): bool
```
Comment (greg0ire) [inline src/Driver/PgSQL/Connection.php:105]: Should these be documented with `@return true`?

File: src/Driver/PgSQL/Result.php [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+{
+    /** @var PgSqlResult|resource|null */
+    private $result;
+
+    /** @param PgSqlResult|resource $result */
+    public function __construct($result)
+    {
+        if (! is_resource($result) && ! $result instanceof PgSqlResult) {
+            throw new TypeError(sprintf(
+                'Expected result to be a resource of an instance of %s, got %s.',
```
Comment (greg0ire) [inline src/Driver/PgSQL/Result.php]:
```suggestion
                'Expected result to be a resource or an instance of %s, got %s.',
```

File: src/Driver/PgSQL/Statement.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * @param PgSqlConnection|resource $connection
+     * @param array<array-key, int>    $parameterMap
+     */
+    public function __construct($connection, string $name, array $parameterMap)
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource of an instance of %s, got %s.',
```
Comment (greg0ire) [inline src/Driver/PgSQL/Statement.php]:
```suggestion
                'Expected connection to be a resource or an instance of %s, got %s.',
```

File: src/Driver/PgSQL/Connection.php (hunk 3 above)
Comment (derrabus) [reply src/Driver/PgSQL/Connection.php:105]: We could do this for most of the non-PDO drivers if we want. On 4.0, we've switched all of them to void.

**Review by greg0ire — DISMISSED** [review]

**Comment by derrabus** [comment]
Running the ORM test suite uncovered a nice little bug in the type mapping logic. 😓 On it.

**Comment by derrabus** [comment]
I've added `Doctrine\DBAL\Tests\Functional\Driver\PgSQL\ResultTest::testTypeConversionWithDuplicateFieldNames()` to reproduce the problem found by the ORM's test suite and fixed the issue.

The ORM's test suite passes now, but it's significantly slower with the new driver than it is with PDO. 😕 

But I can investigate this in a follow-up, I guess.

File: src/Driver/PgSQL/Driver.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+use function pg_result_error;
+use function restore_error_handler;
+use function set_error_handler;
+use function sprintf;
+
+use const PGSQL_CONNECT_FORCE_NEW;
+
+final class Driver extends AbstractPostgreSQLDriver
+{
+    /** @inheritdoc */
```
Comment (greg0ire) [inline src/Driver/PgSQL/Driver.php]: I don't recall the reason but this is usually done like this:

```suggestion
    /** {@inheritDoc} */
```

There are more occurrences in this PR.

File: src/Driver/PgSQL/Driver.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                'dbname' => $params['dbname'] ?? 'postgres',
+                'user' => $params['user'] ?? null,
+                'password' => $params['password'] ?? null,
+                'sslmode' => $params['sslmode'] ?? null,
+            ],
+            static fn ($value) => $value !== '' && $value !== null,
+        );
+
+        return implode(' ', array_map(
+            static fn ($value, string $key) => sprintf('%s=\'%s\'', $key, addslashes($value)),
```
Comment (greg0ire) [inline src/Driver/PgSQL/Driver.php]:
```suggestion
            static fn ($value, string $key) => sprintf("%s='%s'", $key, addslashes($value)),
```

File: src/Driver/PgSQL/Exception/UnknownParameter.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+use function sprintf;
+
+/** @psalm-immutable */
+final class UnknownParameter extends AbstractException
+{
+    public static function new(string $param): self
+    {
+        return new self(
+            sprintf('Could not find parameter %d in the SQL statement', $param),
```
Comment (greg0ire) [inline src/Driver/PgSQL/Exception/UnknownParameter.php]: Should it be `%s` here?

```suggestion
            sprintf('Could not find parameter %s in the SQL statement', $param),
```

File: src/Driver/PgSQL/Driver.php (hunk 6 above)
Comment (derrabus) [reply src/Driver/PgSQL/Driver.php]: Looks like I did this wrong on the SQLite3 driver as well.

File: src/Driver/PgSQL/Exception/UnknownParameter.php (hunk 8 above)
Comment (derrabus) [reply src/Driver/PgSQL/Exception/UnknownParameter.php]: Yes

**Comment by greg0ire** [comment]
> but it's significantly slower with the new driver than it is with PDO

Should we somehow document the new driver as experimental because of this?

**Comment by derrabus** [comment]
Done.

**Review by greg0ire — APPROVED** [review]

**Comment by andrew-demb** [comment]
> The ORM's test suite passes now, but it's significantly slower with the new driver than it is with PDO. 😕

It may be related to "real" prepared statements, in comparison to emulated prepared statements in PDO.

**Comment by derrabus** [comment]
You mean because those real prepared statements mean an extra round-trip to the Posgres server? But that is nevertheless desirable, isn't it? I'm not sure I want to resolve prepared statements in the driver. 🙈
